runhouse/resources/hardware/docker_cluster.py

```diff
@@ -0,0 +1,302 @@
+        this can be an on-demand cluster that is spun up/down through
+        `SkyPilot <https://github.com/skypilot-org/skypilot>`_, using your cloud credentials.
+
+        .. note::
+            To build a cluster, please use the factory method :func:`cluster`.
+        """
+        super().__init__(
+            name=name,
+            ips=[container_name],
+            server_host=server_host or "0.0.0.0",
+            server_port=server_port or DEFAULT_SERVER_PORT,
+            client_port=client_port,
+            den_auth=den_auth,
+            image=image,
+            dryrun=dryrun,
+        )
+        self.server_connection_type = ServerConnectionType.DOCKER
+        self._home_dir = home_dir
+
+    @classmethod
+    def from_name(
+        cls,
+        name: str,
+        load_from_den: bool = True,
+        dryrun: bool = False,
+        _resolve_children: bool = True,
+    ):
+        cluster = super().from_name(
+            name=name,
+            load_from_den=load_from_den,
+            dryrun=dryrun,
+            _resolve_children=_resolve_children,
+        )
+        if cluster and cluster._creds and not dryrun:
+            from runhouse.resources.secrets import Secret
+            from runhouse.resources.secrets.provider_secrets.ssh_secret import SSHSecret
+
+            if isinstance(cluster._creds, SSHSecret):
+                cluster._creds.write()
+            elif isinstance(cluster._creds, Secret):
+                # old version of cluster creds or password only
+                private_key_path = cluster._creds.values.get("ssh_private_key")
+                if private_key_path:
+                    ssh_creds = SSHSecret(
+                        path=private_key_path,
+                        values=cluster._creds.values,
+                    )
+                    ssh_creds.write()
+
+        return cluster
+
+    @property
+    def container_client(self):
+        import docker
+
+        client = docker.from_env()
+
+        return client.containers.get(self.head_ip)
+
+    @property
+    def home_dir(self):
+        if not self._home_dir:
+            # self._home_dir = self.container_client.exec_run("pwd").output.decode().strip()
+            self._home_dir = (
+                self.container_client.exec_run("bash -c 'echo $HOME'")
+                .output.decode()
+                .strip()
+            )
+
+        return self._home_dir
+
+    def config(self, condensed: bool = True):
+        config = super().config(condensed)
+        config["ips"] = self._ips
+        self.save_attrs_to_config(
+            config,
+            [
+                "server_port",
+                "server_host",
+                "server_connection_type",
+                "domain",
+                "den_auth",
+                "ssh_port",
+                "client_port",
+                "ssh_properties",
+            ],
+        )
+
+        if self.image:
+            config["image"] = self.image.config()
```

Comment:
duplicate code from super(), but maybe add `home_dir` / `container` / `server_address` to config